plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/PropertyNameResolver.kt

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2010-2023 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.formver.conversion
+
+import org.jetbrains.kotlin.formver.embeddings.LocalScope
+import org.jetbrains.kotlin.formver.embeddings.ScopedKotlinName
+import org.jetbrains.kotlin.formver.embeddings.SimpleKotlinName
+import org.jetbrains.kotlin.name.Name
+
+/**
+ * Resolver for names of local properties.
+ *
+ * This is a stacked resolver: the resolver for the innermost scope contains a reference
+ * to the resolver for the outer scopes, and automatically searches them.
+ */
+sealed class BasePropertyNameResolver {
+    abstract val parent: BasePropertyNameResolver?
+    abstract fun tryResolveLocalPropertyName(name: Name): ScopedKotlinName?
+    abstract fun registerLocalPropertyName(name: Name)
+
+    fun innerScope(scopeDepth: Int) = ChildPropertyNameResolver(scopeDepth, this)
+}
+
+data object RootPropertyNameResolver : BasePropertyNameResolver() {
+    override val parent: BasePropertyNameResolver? = null
+
+    override fun tryResolveLocalPropertyName(name: Name): ScopedKotlinName? = null
+    override fun registerLocalPropertyName(name: Name) {}
+}
+
+class ChildPropertyNameResolver(private val scopeDepth: Int, override val parent: BasePropertyNameResolver) : BasePropertyNameResolver() {
```

Comment:
with this `parent` logic, can we avoid to pass `scopeDepth` and make it equal to `parent.scopeDepth + 1` ?

Reply:
No, because unfortunately this isn't always the case: let's say `main` calls an inline function `foo` an passes it a lambda `anon`.  We want the scope numbers to be maintained throughout, but each function's scopes are separate and, moreover, name lookup in `anon` does not look at names in `foo` (so we couldn't do it even if we somehow used the same scope stack everywhere).

We could do this if we used the function name for name mangling, but with lambdas it's not super clear how to do that, either.  (Perhaps call depth would work for that.)